[SecuritySolution] Introduce Insights and related alerts by process ancestry

janmonschke opened the pull request:
## Summary

[Research was conducted in 8.3](https://github.com/elastic/security-team/issues/3539#issuecomment-1108757920) in order to identify whether or not it would be possible to find related alerts for any given alert based on the alert's session or it's process ancestry. While this is easily possible for linux-based alerts (based on the session model recently introduced), getting the same information for winlog events prove to be more complicated . The basic related alert for a session for linux was added here: https://github.com/elastic/kibana/pull/127599.

In order to retrieve related alerts for winlog alerts, we're now repurposing the ancestry information utilized by an analyzer endpoint. This endpoint is being worked on in https://github.com/elastic/kibana/pull/135340.

In this PR we're adding an experimental component that will use the aforementioned endpoint to display related alerts and to let analysts investigate those alerts in a new timeline investigation.

In addition to that, these related alerts, alongside other modules that were previously found in the `highlighted fields` section or in the rule description, have now moved into the new `Insights` section as seen on the screenshots. (related cases, related alerts by same source event and related alerts by same session id)

![image](https://user-images.githubusercontent.com/68591/179987014-b40f7c5d-6848-4987-8674-e51341f178d9.png)

![image](https://user-images.githubusercontent.com/68591/179987044-a44b4e20-422e-48c4-8147-86b78fc81b67.png)

### Note

- By default, the module showing related alerts by process ancestry is hidden behind a feature flag since the integration with the endpoint is still pending. You can enable the feature by enabling the feature flag `insightsRelatedAlertsByProcessAncestry`. The module currently uses test data and will most likely not show any alerts in your environment (it will show a count of `3` related alerts but not show any in the table).
- All insights modules, apart from the related alerts by process entry, are loading the data when the flyout is opened. The process ancestry call however can be quite expensive, so it requires a click from the user on the module to start.

### Checklist

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

Closes https://github.com/elastic/kibana/issues/133315
Depends on https://github.com/elastic/kibana/pull/135340

janmonschke committed:
feat: introduce related alerts to highlighted fields

janmonschke committed:
fix: `dataProvider` is `dataProviders` now

janmonschke commented:
@elasticmachine merge upstream

janmonschke commented:
@elasticmachine merge upstream

kibanamachine commented:
merge conflict between base and head

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint --no-cache --fix'

janmonschke committed:
extend `_source` and add `fields` definition

janmonschke committed:
allow to include alert ids when fetching prevalence info

janmonschke committed:
fix: remove broken type import

janmonschke committed:
feat: add insight accordions

janmonschke committed:
feat: show investigate in TL buttons on related-x modules

janmonschke committed:
test: move related_cases tests to new location

janmonschke committed:
chore: remove unused translations

janmonschke committed:
test: remove unused tests

janmonschke committed:
chore: remove unused test data

janmonschke committed:
feat: show limited data note

janmonschke committed:
feat: add `sort` param

janmonschke committed:
test: fix props in test

janmonschke committed:
feat: add click-to-view to process ancestry accordion

janmonschke committed:
test: fix mock import

janmonschke committed:
chore: change order of insight items

janmonschke committed:
chore: use compressed table

janmonschke committed:
chore: update spacing

janmonschke committed:
chore: add empty state to InsightsAccordion

janmonschke committed:
chore: == > ===

janmonschke committed:
test: fix related cases test

janmonschke committed:
chore: comments

janmonschke committed:
test: remove unnecessary `describe`

janmonschke committed:
test: add more tests

janmonschke committed:
fix: remove unused props

janmonschke committed:
chore: simplify the interface of <InsightAccordion />

janmonschke committed:
chore: move translations to their own file

janmonschke committed:
feat: hide related alerts by process ancestry behind feature flag

janmonschke committed:
test: add insights tests

janmonschke commented on x-pack/plugins/security_solution/public/common/containers/alerts/use_alert_prevalence_from_process_tree.ts:
```diff
@@ -0,0 +1,110 @@
+  const query = useQuery<ProcessTreeAlertPrevalenceResponse>(
+    ['getAlertPrevalenceFromProcessTree', processEntityId],
+    () => {
+      return http.get<ProcessTreeAlertPrevalenceResponse>('/TBD', {
```
@kqualters-elastic This would be the module and the place to integrate the endpoint into 🎉

janmonschke committed:
test: add more tests

janmonschke commented:
@elasticmachine merge upstream

michaelolo24 commented on x-pack/plugins/security_solution/common/experimental_features.ts:
```diff
@@ -42,6 +42,11 @@ export const allowedExperimentalValues = Object.freeze({
+  /**
+   * Enables the insights module for related alerts by process ancestry
+   */
+  insightsRelatedAlertsByProcessAncestry: false,
```
Thanks for adding the feature flag. FYI @kqualters-elastic

janmonschke committed:
test: add flyout tests

- Moves the old flyout test to a new file
- Adds a new test case for the related alerts by session use case

michaelolo24 commented on x-pack/plugins/security_solution/cypress/integration/detection_alerts/alert_flyout.spec.ts:
```diff
@@ -0,0 +1,90 @@
+      });
+  });
+
+  it('Opens a new timeline investigation (from an insighrs module)', () => {
```
insights spelled wrong 😛

janmonschke replied:
Nice find :D https://github.com/elastic/kibana/pull/136009/commits/0c3cdcdd99df85fedb7aa71b73019a03b536a9c0

michaelolo24 commented on x-pack/plugins/security_solution/public/common/components/event_details/insights/insight_accordion.tsx:
```diff
@@ -0,0 +1,101 @@
+`;
+
+const EmptyAccordion = euiStyled(StyledAccordion)`
+  background-color: rgba(${({ theme }) => hexToRgb(theme.eui.euiColorDisabled).join(',')}, 0.15);
```
You can add the hex code for 15% transparency to the end. It might be harder for another dev to understand why that's there though, so maybe add a comment with it?

https://gist.github.com/lopspower/03fb1cc0ac9f32ef38f4

janmonschke replied:
Ah, nice one. Will use that next time :)

We decided to remove the background color in the end.

michaelolo24 commented on x-pack/plugins/security_solution/public/common/components/event_details/insights/insight_accordion.tsx:
```diff
@@ -0,0 +1,101 @@
+          />
+        );
+      case 'empty':
+        // Since EuiAccordions don't have an empty state and they don't allow to style the arrow
```
👍🏾

michaelolo24 commented on x-pack/plugins/security_solution/public/common/components/event_details/insights/insights.test.tsx:
```diff
@@ -0,0 +1,75 @@
+  });
+
+  it('renders when there is at least one insight element to show', () => {
+    mockUseGetUserCasesPermissions.mockReturnValue(readCasesPermissions());
```
It's not quickly evident from the way the test is written that there is "something to show". Maybe adding a comment stating cases permissions allows us to display the cases section whether or not there are actually cases to show

janmonschke replied:
Does this explain it? https://github.com/elastic/kibana/pull/136009/commits/0c3cdcdd99df85fedb7aa71b73019a03b536a9c0

:)

michaelolo24 commented on x-pack/plugins/security_solution/public/common/components/event_details/insights/insights.tsx:
```diff
@@ -0,0 +1,127 @@
+    const isRelatedAlertsByProcessAncestryEnabled = useIsExperimentalFeatureEnabled(
+      'insightsRelatedAlertsByProcessAncestry'
+    );
+    const processEntityField = find({ category: 'process', field: 'process.entity_id' }, data);
```
more of an organizational thing, but I think it would be good in the long run to make each insight it's own file. So like cases-insights.tsx could be something like this: It'll make it easier in the long run to reason about I think and make it easier to add additional insights and re-use them.

```
export const useCasesInsights = () => { ... return canShowCasesInsights }

export const CasesInsight = ({ canShowCasesInsights, eventId }) => {
 return canShowCasesInsights ? (
            <EuiFlexItem>
              <RelatedCases eventId={eventId} />
            </EuiFlexItem>
   )
   : null;
}
```

janmonschke replied:
Yeah agreed, not super happy about the rendering duplication :(

michaelolo24 commented on x-pack/plugins/security_solution/public/common/components/event_details/insights/related_alerts_by_process_ancestry.tsx:
```diff
@@ -0,0 +1,173 @@
+/**
+ * Fetches data, displays a loading and error state and notifies about on success
+ */
+const FetchAndNotifyCachedAlertsByProcessAncestry: React.FC<{
```
👌🏾

michaelolo24 commented on x-pack/plugins/security_solution/public/common/components/event_details/insights/simple_alert_table.tsx:
```diff
@@ -0,0 +1,79 @@
+  },
+];
+
+const alertLimit = 10;
```
Can you add a note why it's limited to 10 in case there's a desire to change it in the future?

janmonschke replied:
Sure thing: https://github.com/elastic/kibana/pull/136009/commits/0c3cdcdd99df85fedb7aa71b73019a03b536a9c0

michaelolo24 commented on x-pack/plugins/security_solution/public/common/components/event_details/insights/translations.ts:
```diff
@@ -0,0 +1,116 @@
+});
+
+export const PROCESS_ANCESTRY = i18n.translate(
+  'xpack.securitySolution.alertDetails.overview.insights.related_alerts_by_process_ancestry',
```
random empty space?

janmonschke replied:
Prettier moved it to the next line as the line got quite long with that id

michaelolo24 reviewed (approved):
Just a few comments, but nothing too crazy. Pulled down and tested locally in the alert page, timeline view, and cases pages. All looks good and consistent. Great job with these changes! 🚀 LGTM!

janmonschke committed:
chore: comments

andrew-goldstein commented on x-pack/plugins/security_solution/public/common/components/event_details/insights/insight_accordion.tsx:
```diff
@@ -0,0 +1,101 @@
+}
+
+/**
+ * A special according that is used in the Insights section on the alert flyout.
```
nit: `s/according/accordion`

andrew-goldstein commented on x-pack/plugins/security_solution/public/common/containers/alerts/use_alert_prevalence_from_process_tree.ts:
```diff
@@ -0,0 +1,110 @@
+import { useQuery } from 'react-query';
+import { useHttp } from '../../lib/kibana';
+
+// import { DEFAULT_MAX_TABLE_QUERY_SIZE } from '../../../../common/constants';
```
nit: these may be deleted

andrew-goldstein replied:
update: if this is commented-out example code for when the `/TBD` endpoint exists, please ignore this comment

janmonschke replied:
Yep, that is what this was for 👍

andrew-goldstein commented on x-pack/plugins/security_solution/public/common/containers/alerts/use_alert_prevalence_from_process_tree.ts:
```diff
@@ -0,0 +1,110 @@
+  timelineId,
+  signalIndexName,
+}: UseAlertPrevalenceOptions): UserAlertPrevalenceFromProcessTreeResult => {
+  // const timelineTime = useDeepEqualSelector((state) =>
```
nit: these may be removed

andrew-goldstein replied:
(if not intended as sample code for the `/TBD` endpoint)

andrew-goldstein reviewed (approved):
Thanks for this feature @janmonschke! 🙏 
Desk tested locally
LGTM 🚀

janmonschke commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into security/related-alerts-by-process-ancestry

janmonschke committed:
fix: typo

janmonschke committed:
feat: remove background color in empty state

MadameSheema commented on x-pack/plugins/security_solution/cypress/integration/detection_alerts/alert_flyout.spec.ts:
```diff
@@ -0,0 +1,90 @@
+      .invoke('text')
+      .then((alertCount) => {
+        // Click on the first button that lets us investigate in timeline
+        cy.get(ALERT_FLYOUT).find(SUMMARY_VIEW_INVESTIGATE_IN_TIMELINE_BUTTON).first().click();
```
nit: Instead of adding comments, you can wrap that code into meaningful methods and add them to the `tasks` folder. In this way, everyone can easily reuse the code in the future if needed.

janmonschke committed:
test: the `auditbeat` archive is loaded by default

- also: removing the `auditbeat` archive can lead to unwanted side-effects (TIL)

janmonschke commented on x-pack/plugins/security_solution/cypress/integration/detection_alerts/alert_flyout.spec.ts:
```diff
@@ -0,0 +1,90 @@
+describe('Alert Flyout', () => {
+  before(() => {
+    cleanKibana();
+    esArchiverLoad('auditbeat');
```
Nice catch, I removed it here https://github.com/elastic/kibana/pull/136009/commits/e3829de927dcb4bdeb19c65c6a7980b4d598fbb7 🎉

MadameSheema commented on x-pack/plugins/security_solution/cypress/integration/detection_alerts/alert_flyout.spec.ts:
```diff
@@ -0,0 +1,90 @@
+  });
+
+  afterEach(() => {
+    cy.get(CLOSE_TIMELINE_BUTTON).click();
```
nit: you can reuse the `closeTimeline` method placed in `x-pack/plugins/security_solution/cypress/tasks/timeline.ts`. Moreover, the method was written in a way to avoid flakiness in certain situations.

MadameSheema commented on x-pack/test/security_solution_cypress/es_archives/auditbeat/data.json:
```diff
@@ -83,7 +83,12 @@
           "-zsh"
         ],
-        "entity_id" : "q6pltOhTWlQx3BCD"
+        "entity_id" : "q6pltOhTWlQx3BCD",
```
Really good approach adding the data needed instead of creating a new file. Thanks! :)

janmonschke committed:
test: refactor cypress test

- use `closeTimeline()`
- move `openOverview()` to beforeEach

MadameSheema reviewed (approved):
sec-eng-prod changes LGTM, thanks! :)

janmonschke committed:
test: remove unnecessary call to `openOverview()`

janmonschke commented on x-pack/plugins/security_solution/cypress/integration/detection_alerts/alert_flyout.spec.ts:
```diff
@@ -0,0 +1,90 @@
+  });
+
+  it('Opens a new timeline investigation (from a prevalence field)', () => {
+    openOverview();
```
As discussed offline, we were able to remove the method in the end :) https://github.com/elastic/kibana/pull/136009/commits/e14f8f0a56b6c03d84e6101c11405d5ebe43f40b

janmonschke committed:
fix: make sure related cases has a loading state

janmonschke committed:
chore: remove unused file

janmonschke committed:
chore: remove unused translations

janmonschke committed:
chore: remove unused translations

YulNaumenko reviewed (approved):
LGTM!

marshallmain reviewed (approved):
Codeowner review on `query_signals_index_schema` and `query_signals_route` only LGTM
